src/main/scala/workflow/graph/EquivalentNodeMergeRule.scala

```diff
@@ -0,0 +1,47 @@
+package workflow.graph
+
+/**
+ * A rule to merge equivalent nodes in the DAG.
+ * Nodes are considered equivalent if:
+ * - The operators stored within the nodes are equal (.equals() is true)
+ * - They share the same dependencies
+ *
+ * This rule also merges prefixes if any of
+ * the nodes being merged have their prefix attached.
+ */
+object EquivalentNodeMergeRule extends Rule {
+  override def apply(plan: Graph, prefixes: Map[NodeId, Prefix]): (Graph, Map[NodeId, Prefix]) = {
+    val nodeSetsToMerge = plan.nodes.groupBy(id => (plan.getOperator(id), plan.getDependencies(id))).values
```

Comment:
`groupBy` builds a hash map in scala if I'm not wrong and hash maps are grouped by calling `.hashCode` and not `.equals` -- So we should either change the doc string above to say we use `.hashCode` (does every operator have one ?) or change the code to explicitly use `.equals` ?